File: Translated RPi Code/main/message.cpp
#include "Message.h"

// Constructor
Message::Message(const String& message, int x_pos, int y_pos, int rot, float horz_offset, float vert_offset, int scroll_time, Adafruit_SSD1309* oled)
    : message(message), x_pos(x_pos), y_pos(y_pos), rot(rot), horz_offset(horz_offset), vert_offset(vert_offset), scroll_time(scroll_time), oled(oled) {
    
    scrollable = false;
    start_time = millis();
    message_len = measureTextWidth(message);
}

// Sets the message and updates its length
void Message::setMessage(const String& m) {
    message = m;
    message_len = measureTextWidth(m);
}

// Determines if the message is scrollable
void Message::determineScrollable() {
    // If the message does not fit on the screen, make it scrollable
    if (message_len > oled->height()) {
        scrollable = true;

        // Pad the message with spaces for centering
        int num_spaces = message.length() / 2;
        for (int i = 0; i < num_spaces; i++) {
            message = " " + message;
        }

        // Update the message length
        message_len = measureTextWidth(message);
    } else {
        scrollable = false;
    }
}

// Draws the message and updates scrolling if necessary
void Message::drawMessage() {
    // Handle scrolling
    if (scrollable) {
        unsigned long current_time = millis();
        if (current_time - start_time >= scroll_time * 1000) {
            // Scroll by moving the first character to the end
            message = message.substring(1) + message[0];
            start_time = current_time;
        }
    }

    // Draw the message on the display at the specified position
    int text_x = x_pos + round(message_len * horz_offset);
    int text_y = y_pos + round(message_len * vert_offset);
    oled->setRotation(rot);
    oled->setCursor(text_x, text_y);
    oled->print(message);
    oled->display();
}

// Approximates the width of the text based on character width (adjust as needed)
int Message::measureTextWidth(const String& text) {
    int char_width = 5;  // Approximate width per character for a 5x8 font
    return text.length() * char_width;
}

File: Backup Pi Code/message.py
"""
Victor Cappuzzo

Class for a Message object
"""

from ssd1309 import Display
from xglcd_font import XglcdFont
import time

class Message():

    # Initialization
    def __init__(self, message, x_pos, y_pos, font, rot, horz_offset, vert_offset, scroll_time, oled):
        
        self.message = message
        self.x_pos = x_pos
        self.y_pos = y_pos
        self.font = font
        self.rot = rot
        self.horz_offset = horz_offset
        self.vert_offset = vert_offset
        self.scroll_time = scroll_time
        self.oled = oled

        self.scrollable = False
        self.start_time = time.time()
        self.message_len = self.font.measure_text(self.message)
    
    # Sets the message of the object and updates its length
    def setMessage(self, m):
        
        self.message = m
        self.message_len = self.font.measure_text(m)
    
    # Determine if the message is scrollable (does not fit in the bounds of the screen)
    def determineScrollable(self):
        
        # If the message does not fit on the screen, make it scrollable
        if self.message_len > self.oled.height:
            
            self.scrollable = True

            # Pad the message at the beginning with spaces so that the message starts in the middle of the screen
            num_spaces = len(self.message) // 2
            for _ in range(num_spaces):
                self.message = " " + self.message

            # Update the message length
            self.message_len = self.font.measure_text(self.message)
        
        # If the message does fit on the screen, scrolling is not necessary
        else:

            self.scrollable = False

    # Draw the message (if it is scrollable, update the messages movement)
    def drawMessage(self):
        
        # If the message is scrollable
        if self.scrollable:

            # Get the current timer reading
            current_time = time.time()

            # Check if the timer has surpassed the scroll time
            if current_time - self.start_time >= self.scroll_time:

                # Take a character from the beginning of the message and move it to the end
                self.message = self.message[1: len(self.message)] + self.message[0]

                # Reset the start timer to the current time (keeps track of elapsed time)
                self.start_time = current_time

        
        # Draw the message
        self.oled.draw_text(self.x_pos + round(self.message_len*self.horz_offset), self.y_pos + round(self.message_len*self.vert_offset), self.message, self.font, rotate=self.rot)
        #self.oled.present()
